Document `dart pub token`. (#3693)
dart-lang/site-www: src/tools/pub/cmd/pub-token.md

---
title: dart pub token
description: Manage authentication tokens for hosted pub repositories.
---

_pub_ _token_ is one of the subcommands of the [pub command](/tools/pub/cmd).

It is used to update a store of credential tokens used for authenticating
against third-party servers when [publishing](pub-lish) a package and when
[retrieving](pub-get) dependencies.

It has three subcommands: `add`, `list` and `remove`.

> If you try to `pub get` and have a [dependency](/tools/pub/dependencies) hosted

Reviewer: ```suggestion
If you try to `dart pub get` and have a [dependency](/tools/pub/dependencies) hosted
```

on a private repository you will be asked to provide credentials:
```terminal
$ dart pub get
Resolving dependencies... 
https://some-package-server.com/my-org/my-repo package repository requested authentication! You can provide credential using:
    pub token add https://some-package-server.com/my-org/my-repo

Go to https://some-package-server.com and log in to obtain your token. 
```

> The last line is a message the server can provide to help you obtaining a token.

Reviewer: Maybe omit that here... and omit the last line..

-------------------

Or clarify that some servers may not provide such help message.

---
title: dart pub token
description: Manage authentication tokens for package repositories.
---

_pub_ _token_ is one of the subcommands of the [pub command](/tools/pub/cmd).

It is used to update a store of credential tokens used for authenticating
against third-party servers when [publishing](pub-lish) a package and when
[retrieving](pub-get) dependencies.

It has three subcommands: `add`, `list` and `remove`.

If you try to `dart pub get` and have a [dependency](/tools/pub/dependencies) hosted
on a private repository you will be asked to provide credentials:

> ```terminal

Reviewer 2: ```suggestion

```terminal
```

$ dart pub get
Resolving dependencies... 
https://some-package-repo.com/my-org/my-repo package repository requested authentication! You can provide credential using:
    pub token add https://some-package-repo.com/my-org/my-repo

Go to https://some-package-repo.com and log in to obtain your token. 
```

The last line is a message the server can provide to help you obtaining a token.
Some servers might not provide such a message.

> To enter the credentials use the `add` command, and type the credential on
> stdin.

Reviewer 2: ```suggestion
To enter the credentials use the `add` subcommand, 
and type the credential on stdin.
```

---
title: dart pub token
description: Manage authentication tokens for package repositories.
---

_pub_ _token_ is one of the subcommands of the [pub command](/tools/pub/cmd).

> It is used to update a store of credential tokens used for authenticating
> against third-party servers when [publishing](pub-lish) a package and when
> [retrieving](pub-get) dependencies.

Reviewer: ```suggestion
It is used to manage a store of secret tokens for authenticating
against third-party servers when [publishing](pub-lish) packages and
[retrieving](pub-get) dependencies.
```

> The token store is stored in a

Reviewer: ```suggestion
The tokens are stored in a
```

[user-wide config dir](https://github.com/dart-lang/cli_util/blob/71ba36e2554f7b7717f3f12b5ddd33751a4e3ddd/lib/cli_util.dart#L88-L118). 

It has three subcommands: `add`, `list` and `remove`.

If you try to `dart pub get` and have a [dependency](/tools/pub/dependencies) hosted

> on a private repository you will be asked to provide credentials:

Reviewer: ```suggestion
on a private repository you _may_ be asked to provide credentials:
```

```terminal
$ dart pub get
Resolving dependencies... 
https://some-package-repo.com/my-org/my-repo package repository requested authentication! You can provide credential using:
    pub token add https://some-package-repo.com/my-org/my-repo

Go to https://some-package-repo.com and log in to obtain your token. 
```

The last line is a message the server can provide to help you obtaining a token.
Some servers might not provide such a message.

## Adding credentials `dart pub token add`

> To enter the credentials use the `add` subcommand, 

Reviewer: I wonder if we should spell out the entire command, like `dart pub token add`

Author: Done

and type the credential on stdin.

```terminal
$ dart pub token add https://some-package-repo.com/my-org/my-repo
Enter secret token: <Type token on stdin>
 Requests to "https://some-package-repo.com/my-org/my-repo" will now be 
 authenticated using the secret token.
```

In a scripting situation you can pipe the credential to `add`. This might leave
a trace in the command history if you are not careful.

```terminal

> $ echo <credential> | dart pub token add https://other-package-repo.com/

Reviewer: ```suggestion
$ echo '<credential>' | dart pub token add https://other-package-repo.com/
```

```

> 

Reviewer: In scripting situations where the token is stored in an environment variable like `MY_SECRET_TOKEN`

You can also do: `dart pub token add <hosted-url> --env-var MY_SECRET_TOKEN`

This will cause `dart pub get` to read whatever is stored in `MY_SECRET_TOKEN` and use it as the authentication token.

Author: Yeah - let us suggest that instead.

on a private repository you _may_ be asked to provide credentials:

```terminal
$ dart pub get
Resolving dependencies... 
https://some-package-repo.com/my-org/my-repo package repository requested authentication! You can provide credential using:
    pub token add https://some-package-repo.com/my-org/my-repo

Go to https://some-package-repo.com and log in to obtain your token. 
```

The last line is a message the server can provide to help you obtaining a token.
Some servers might not provide such a message.

## Adding credentials `dart pub token add`

To enter the credentials use `dart pub token add`, 
and type the credential on stdin.

```terminal
$ dart pub token add https://some-package-repo.com/my-org/my-repo
Enter secret token: <Type token on stdin>
 Requests to "https://some-package-repo.com/my-org/my-repo" will now be 
 authenticated using the secret token.
```

{{site.alert.note}}
  The token is input on stdin rather than as a command line option to avoid it
  ending up in the shell history such as `~/.bash_history`.
{{site.alert.end}}

In a scripting situation you can store the secret in an environment variable and

>  use `dart pub token add --env-var`.

Reviewer: ```suggestion
 use `dart pub token add <hosted-url> --env-var <ENV_VAR_NAME>`.
```

> ```terminal
> $ dart pub token add https://other-package-repo.com/ --env-var TOKEN_VAR
> Requests to "https://other-package-repo.com/" will now be authenticated using the secret token stored in the environment variable "TOKEN_VAR".

Reviewer: ```suggestion
```terminal
export TOKEN_VAR='<secret token>'
$ dart pub token add https://other-package-repo.com/ --env-var TOKEN_VAR
Requests to "https://other-package-repo.com/" will now be authenticated using the secret token stored in the environment variable "TOKEN_VAR".
```

Author: I instead explain how to use the secrets functionality of known CI systems.
